e3sm_diags/run.py

```diff
@@ -18,47 +20,120 @@ class Run:
+              * If False, run all sets using the list of parameters passed in this
+                function and parameters defined in a .cfg file (if defined).
+                This is the default option.
+
+        Returns
+        -------
+        List[CoreParameter]
+            A list of parameter objects with their results.
+
+        Raises
+        ------
+        RuntimeError
+            If a diagnostic run using a parameter fails for any reason.
         """
-        Based on sets_to_run, run the diags with the list of parameters.
-        """
-        final_params = self.get_final_parameters(parameters)
-        if not final_params:
-            msg = "No parameters we able to be extracted."
-            msg += " Please check the parameters you defined."
-            raise RuntimeError(msg)
+
+        params = self.get_final_parameters(parameters, debug)
+
+        if params is None or len(params) == 0:
+            raise RuntimeError(
+                "No parameters we able to be extracted. Please "
+                "check the parameters you defined."
+            )
 
         try:
-            main(final_params)
+            params_results = main(params)
         except Exception:
             logger.exception("Error traceback:", exc_info=True)
-        move_log_to_prov_dir(final_params[0].results_dir)
 
-    def get_final_parameters(self, parameters):
+        move_log_to_prov_dir(params_results[0].results_dir)
+
+        return params_results
+
+    def get_final_parameters(
+        self, parameters: List[CoreParameter], debug: bool = False
+    ):
         """
         Based on sets_to_run and the list of parameters,
         get the final list of paremeters to run the diags on.
         """
-        if not parameters or not isinstance(parameters, list):
-            msg = "You must pass in a list of parameter objects."
-            raise RuntimeError(msg)
+        self._validate_parameters(parameters)
+
+        # FIXME: This line produces some unintended side-effects. For example,
+        # let's say we have two objects: 1. CoreParameter, 2. ZonalMean2DParameter.
+        # If object 1 has `plevs=[200]`, this value will get copied to object 2.
+        # Object 2 has a check to make sure plevs has more than 1 value. This
+        # breaks the diagnostic run as a result. The workaround is to loop
+        # over `run_diags()` function and run one parameter at a time.
+        self._add_parent_attrs_to_children(parameters)
 
-        # For each of the passed in parameters, we can only have one of
-        # each type.
-        types = set([p.__class__ for p in parameters])
-        if len(types) != len(parameters):
-            msg = "You passed in two or more parameters of the same type."
-            raise RuntimeError(msg)
+        if debug:
+            final_params = self._get_debug_parameters(parameters)
+        else:
+            final_params = self._get_run_parameters(parameters)
```

Comment:
We now have the option to run in `debug` mode, which means only the parameters explicitly passed via the Python script are used. Default diagnostics are no longer ran because they pollute the console output and make it hard to debug just the specific sets we care about.